### PR TITLE
[processing] Add r.geomorphon grass command and module in grass core plugin

### PR DESCRIPTION
**Pull request by epifanio**
## Description
This PR will enable the module ´r.geomorphon´ in the list of available ´QGIS-GRASS´ modules.
It is a recently added module for terrain analysis. It has been merged into grass starting from version 7.6  

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [x] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [x] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [ ] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and contain sufficient information in the commit message to be documented
- [x] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [x] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit

### CODE REVIEW

**Comment by luipir** [comment]
@epifanio would be fin add a test (follow as int he other grass tests)... a simple test not for all parameters. It is useful just to test that the algorithm is correctly managed by processing.

File: src/plugins/grass/modules/r.geomorphon.qgm [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE qgisgrassmodule SYSTEM "http://mrcc.com/qgisgrassmodule.dtd">
+
+<qgisgrassmodule label="Calculates geomorphons (terrain forms) and associated geometry using machine vision approach" module="r.geomorphon">
+	<option key="elevation" answer="" hidden="no"/>
```
Comment (luipir) [inline src/plugins/grass/modules/r.geomorphon.qgm]: are there much more parameters that that here exposed?

**Comment by nyalldawson** [comment]
@luipir 

Actually this PR isn't related to processing - it's only the standalone grass plugin which it impacts.

@epifanio The more modern way of running these scripts is through Processing. Are you able to also add your changes to the Processing GRASS provider so that the new tools can be run through the Processing toolbox?

**Comment by luipir** [comment]
@nyalldawson gosh! that's the reason I wasn't able to recgnise the syntax... but I suppose was my problem. Is this plougin still in use and maintained? @epifanio I strongly suggest to move your PR to processing.
Look in https://github.com/qgis/QGIS/blob/master/python/plugins/processing/algs/grass7/grass7.txt
and add alg description in https://github.com/qgis/QGIS/tree/master/python/plugins/processing/algs/grass7/description
and related test as in:
https://github.com/qgis/QGIS/tree/master/python/plugins/processing/tests

File: src/plugins/grass/modules/r.geomorphon.qgm (hunk 1 above)
Comment (epifanio) [reply src/plugins/grass/modules/r.geomorphon.qgm]: Those are the main parameters, enough to make the module usabble and keep its usage simpler. I can think of adding the extra output parameters and the multiresolution marked as advanced.

**Comment by epifanio** [comment]
> @nyalldawson gosh! that's the reason I wasn't able to recgnise the syntax... but I suppose was my problem. Is this plougin still in use and maintained? 

You managed some how .. to let me feel old ... ;) ... I'll loock into the 'processing way' to do it ;)

**Comment by epifanio** [comment]
I added the description file into the processing folder and added the relative test in the yaml file.

**Comment by epifanio** [comment]
@luipir  what is the exact command to generate the rasterhash? 
I had this error in the test:
```
AssertionError: 'e1a1fc36d53e2ac801632f9e2271922d2b1340869f2db82d5edaeb0e' != '0a35ad47518c3a5dd66752c4a6d6e87e'
```
My best guess is that the hash type I used to generate the `md5sum:0a35ad47518c3a5dd66752c4a6d6e87e` is not the same used by QGIS test framework.

Any clue on how to generate a proper rasterhash?

**Comment by luipir** [comment]
> @luipir what is the exact command to generate the rasterhash?

@epifanio did you check how is done in processing test? or beter did you followed the "common" way to generate test dataset? 
https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/README.md#raster-files

**Comment by epifanio** [comment]
> > @luipir what is the exact command to generate the rasterhash?
> 
> @epifanio did you check how is done in processing test? or better did you followed the "common" way to generate test dataset?
> https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/README.md#raster-files

This is what I added in the `yaml` test file:

```
  - algorithm: grass7:r.geomorphon
    name: GRASS7 r.geomorphon
    params:
      GRASS_REGION_CELLSIZE_PARAMETER: 0.0
      GRASS_REGION_PARAMETER: 344500.0,358400.0,6682800.0,6693700.0
      elevation:
        name: custom/grass7/float_raster.tif
        type: raster
      dist: '0'
      flat: 1.0
      search: '3'
      skip: 0
    results:
      forms:
        hash: 0a35ad47518c3a5dd66752c4a6d6e87e
        type: rasterhash
```
the corresponding `GRASS ` command is:

```
r.geomorphon dist=0 flat=1 search=3 skip=0 elevation=float_raster.tif forms=test_geomorphon
```

I then saved the output and run a series of tools to generate the hash code on it (my previous comment).. but no one of the ash code  I generated matched the one used by the testing tool. So, I'm wondering how to generate the proper hash code and if I'm pointing it to the right file

**Comment by luipir** [comment]
@epifanio as said in: https://github.com/qgis/QGIS/blob/master/python/plugins/processing/tests/README.md#raster-files says
"This is calculated when you create a test from the processing history." 

as from ypou answer you did all manually not following the process... try following the automatic process to generate tes... there is a guide for this

**Review by luipir — APPROVED** [review]
great, tnx @epifanio

**Comment by luipir** [comment]
@nyalldawson IMHO safe to merge also with modification to core grass plugin  src/plugins/grass/modules. Do you agree?

**Review by m-kuhn — APPROVED** [review]
I have just removed some leftover files and it should be good to ship now.

Just remember to squash on merge

Thanks @epifanio

**Comment by luipir** [comment]
> have just removed some leftover files and it should be good to ship now.

tnx @m-kuhn

**Comment by epifanio** [comment]
Thanks @luipir @nyalldawson @m-kuhn

**Pull request merged**